**Problem 1.** Define fair as having an expected value of $0.  
Probability of rolling a 7 is 6/36 = 1/6.  

If you win, the payoff is +4 chips. If you lose, the payoff is –1 chip.  

Expected value = (1/6 * 4) + (5/6 * –1)  
= (4/6) – (5/6)  
= –1/6  

Not fair. Fair would be with 5 chips instead of 4 for a win.


In [3]:
import random

times = 1000000
money = 0
for i in range(times):
    d1 = random.randint(1, 6)
    d2 = random.randint(1, 6)
    s = d1 + d2
    if s == 7:
        money += 4
    else:
        money -= 1
print("Average return per bet:", money / times)

Average return per bet: -0.16837


The simulation average is about -0.168, matching EV=-1/6, so the game isn't fair. A +5 win would make it fair.

**Problem 2 (Counting).** You roll three fair dice. What is the probability that the sum is not a prime number?

Total outcomes: \(6^3 = 216\). Prime sums in 3..18 are 3, 5, 7, 11, 13, 17.

Counts of three-dice sums:
3:1, 4:3, 5:6, 6:10, 7:15, 8:21, 9:25, 10:27, 11:27, 12:25, 13:21, 14:15, 15:10, 16:6, 17:3, 18:1.

Prime-sum count = \(1 + 6 + 15 + 27 + 21 + 3 = 73\)

Non-prime outcomes = \(216 - 73 = 143\)

In [4]:
count = 0
for d1 in range(1, 7):
    for d2 in range(1, 7):
        for d3 in range(1, 7):
            s = d1 + d2 + d3
            if s not in [3, 5, 7, 11, 13, 17]:
                count += 1
            else:
                pass
print(count / (6 * 6 * 6))

0.6620370370370371


The simulation is close to 143/216 ≈ 0.6620, confirming the counting result.


In [7]:
n = 100
lockers = [False] * (n + 1)
for student in range(1, n + 1):
    for locker in range(student, n + 1, student):
        lockers[locker] = not lockers[locker]
    line = ""
    for i in range(1, n + 1):
        line += "* " if lockers[i] else "  "
    print(line)
open_lockers = [i for i in range(1, n + 1) if lockers[i]]
print("\nOpen lockers at the end:")
print(open_lockers)

* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * 
*   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   *   
*       * * *       * * *       * * *       * * *       * * *       * * *       * * *       * * *       * * *       * * *       * * *       * * *       * * *       * * *       * * *       * * *       
*     * * * * *     *   *     * * * * *     *   *     * * * * *     *   *     * * * * *     *   *     * * * * *     *   *     * * * * *     *   *     * * * * *     *   *     * * * * *     *   *     * 
*     *   * * *   * *   *   * * * * *       *         * *   * *         *       * * * * *   *   * *   * * *   *     * * *     *   * * *   * *   *   * * * * *       *         * *   * *         *   

The lockers left open are exactly the perfect squares: 1, 4, 9, 16, 25, 36, 49, 64, 81, 100.

This is because locker k is toggled once for each of its divisors. Non square numbers have divisors in pairs, so they are toggled an even number of times and end closed. On the other hand, Perfect squares end open because one of their divisors is repeated.“*” means the lockers are open.
